File: Tracker/src/main/java/com/cyrillrx/tracker/event/LogEvent.java
package com.cyrillrx.tracker.event;

import com.cyrillrx.tracker.context.TrackerContext;

import java.util.Map;

/**
 * @author Cyril Leroux
 *         Created on 25/04/2016.
 */
public class LogEvent extends TrackEvent {

    protected int severity;
    protected String tag;
    protected String message;
    protected String screen;
    protected String screenExecId;

    protected LogEvent() { }

    public int getSeverity() { return severity; }

    public String getTag() { return tag; }

    public String getMessage() { return message; }

    public String getScreen() { return screen; }

    public String getScreenExecId() { return screenExecId; }

    public static class Builder extends EventBuilder<LogEvent> {

        public Builder() { event = new LogEvent(); }

        public LogEvent build() {

            if (event.category == null || event.category.isEmpty()) {
                throw new IllegalStateException("Category is mandatory");
            }

            return event;
        }

        @Override
        public Builder setContext(TrackerContext context) { return (Builder) super.setContext(context); }

        @Override
        public Builder setCategory(String category) { return (Builder) super.setCategory(category); }

        @Override
        public Builder setName(String name) { return (Builder) super.setName(name); }

        @Override
        public Builder setSource(String source) { return (Builder) super.setSource(source); }

        @Override
        public Builder putCustomAttribute(String key, String value) {
            return (Builder) super.putCustomAttribute(key, value);
        }

        @Override
        public Builder putCustomAttributes(Map<String, String> values) {
            return (Builder) super.putCustomAttributes(values);
        }

        @Deprecated
        @Override
        public Builder setId(String id) { return (Builder) super.setId(id); }

        @Deprecated
        @Override
        public Builder setType(String type) { return (Builder) super.setType(type); }

        public Builder setSeverity(int severity) {
            event.severity = severity;
            return this;
        }

        public Builder setTag(String tag) {
            event.tag = tag;
            return this;
        }

        public Builder setMessage(String message) {
            event.message = message;
            return this;
        }

        public Builder setScreen(String screen) {
            event.screen = screen;
            return this;
        }

        public Builder setScreenExecId(String screenExecId) {
            event.screenExecId = screenExecId;
            return this;
        }
    }
}

File: tracker/src/main/java/com/cyrillrx/tracker/event/LogEvent.kt
package com.cyrillrx.tracker.event

import com.cyrillrx.tracker.context.TrackerContext
import com.cyrillrx.tracker.utils.requireNotBlank
import java.util.HashMap

/**
 * @author Cyril Leroux
 *         Created on 25/04/2016.
 */
class LogEvent protected constructor(
    /** Global context. Not directly related to the event.  */
    val context: TrackerContext,
    val category: String,
    val name: String,
    /** Source of the event (class name, or ui screen)  */
    val source: String? = null
) {

    var severity = 0
        protected set
    var tag: String? = null
        protected set
    var message: String? = null
        protected set
    var screen: String? = null
        protected set
    var screenExecId: String? = null
        protected set

    class Builder {
        private var context: TrackerContext? = null
        private var category: String? = null
        private var name: String? = null
        /** Source of the event (class name, or ui screen)  */
        private var source: String? = null

        private val customAttributes: HashMap<String, Any> = HashMap()

        fun build(): LogEvent {
            val context = requireNotNull(context) { "TrackerContext is mandatory. Was null." }

            val category = requireNotBlank(category) { "Category is mandatory." }

            val name = requireNotBlank(name) { "Event name is mandatory." }

            return LogEvent(context, category, name, source)
        }

        fun setContext(context: TrackerContext) = apply {
            this.context = context
        }

        fun setCategory(category: String) = apply {
            this.category = category
        }

        fun setName(name: String) = apply {
            this.name = name
        }

        fun setSource(source: String) = apply {
            this.source = source
        }

        fun putCustomAttribute(key: String, value: Any) = apply {
            this.customAttributes[key] = value
        }

        fun <T : Any> putCustomAttributes(values: Map<String, T>) = apply {
            this.customAttributes.putAll(values)
        }
    }
}